### PR TITLE
Tweak errors coming from `for`-loop, `?` and `.await` desugaring

### LINKED ISSUE #66731: More friendly error msg when await on NONE ASYNC fn/block or return a obj that implements deprecated Future
```rust
fn bar () -> impl futures::future::Future<Item=(), Error=()> {}

fn boo (){}

async fn foo() -> std::io::Result<()> {
    boo().await;
    bar().await;
    std::io::Result::Ok(())
}
```
```rust
error[E0277]: the trait bound `(): std::future::Future` is not satisfied
   --> src/main.rs:6:5
    |
6   |     boo().await;
    |     ^^^^^^^^^^^ the trait `std::future::Future` is not implemented for `()`

error[E0277]: the trait bound `impl futures::future::Future: std::future::Future` is not satisfied
   --> src/main.rs:7:5
    |
7   |     bar().await;
    |     ^^^^^^^^^^^ the trait `std::future::Future` is not implemented for `impl futures::future::Future`

error: aborting due to 2 previous errors
```
1. First error should more like`await need a async function, async block or return a type that impl std::future::Future, did you forget add async before fn boo() ?`
2. Second error should more like `You are return a Future that has been droped support or deprecated, maybe try to upgrade to latest version of futures crate?`
3. if 3rd party crate depends on derpecated crate that uses deprecated Future trait, the compiler should also tell use explicitly by some error message. (e.g.  reqwest 0.9.22 depends on futures = "0.1.23")

I think more friendly error message like these will help new comers from javascript -> front end devs to more quickly solve problem, because they don't have multiple `Promise` implements, current error messages will make them very confusing and frustrating

### PR DESCRIPTION
**Pull request by estebank**
* Suggest removal of `.await` on non-`Future` expression
 * Keep track of obligations introduced by desugaring
 * Remove span pointing at method for obligation errors coming from desugaring
 * Point at called local sync `fn` and suggest making it `async`

```
error[E0277]: `()` is not a future
  --> $DIR/unnecessary-await.rs:9:10
   |
LL |     boo().await;
   |     -----^^^^^^ `()` is not a future
   |     |
   |     this call returns `()`
   |
   = help: the trait `Future` is not implemented for `()`
help: do not `.await` the expression
   |
LL -     boo().await;
LL +     boo();
   | 
help: alternatively, consider making `fn boo` asynchronous
   |
LL | async fn boo () {}
   | +++++
```

Fix #66731.

### CODE REVIEW

File: src/test/ui/async-await/issues/issue-65436-raw-ptr-not-send.stderr [hunk 1]
```diff
@@ -6,16 +6,12 @@ LL |     assert_send(async {
...
    |
 LL |         bar(Foo(std::ptr::null())).await;
-   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ first, await occurs here, with `std::ptr::null()` maybe used later...
-note: `std::ptr::null()` is later dropped here
-  --> $DIR/issue-65436-raw-ptr-not-send.rs:14:41
-   |
-LL |         bar(Foo(std::ptr::null())).await;
-   |                 ----------------        ^
-   |                 |
+   |                 ----------------  ^^^^^^- `std::ptr::null()` is later dropped here
```
Comment (tmandry) [inline src/test/ui/async-await/issues/issue-65436-raw-ptr-not-send.stderr:12]: This message being first is confusing, is there a way to control the order or reword it to be less confusing?

File: src/test/ui/async-await/unnecessary-await.rs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+// edition:2018
+
+async fn foo () { }
+fn bar () -> impl std::future::Future { async {} }
+fn boo () {}
```
Comment (tmandry) [inline src/test/ui/async-await/unnecessary-await.rs:5]: nit: shouldn't be spaces before argument lists :)

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs [hunk 3]
```diff
@@ -873,6 +879,61 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
...
+        obligation: &PredicateObligation<'tcx>,
+        err: &mut DiagnosticBuilder<'_>,
+    ) {
+        let span = obligation.cause.span;
+
+        if let ObligationCauseCode::AwaitableExpr(hir_id) = obligation.cause.code {
+            // FIXME: use `obligation.predicate.kind()...trait_ref.self_ty()` to see if we have `()`
+            // and if not maybe suggest doing something else? If we kept the expression around we
+            // could also check if it is an fn call (very likely) and suggest changing *that*, if
+            // it is from the local crate.
```
Comment (tmandry) [inline compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:893]: Looks like at least some of this is done now, am I misunderstanding?

File: library/core/src/ops/try_trait.rs [hunk 4]
```diff
@@ -115,15 +115,14 @@ use crate::ops::ControlFlow;
 #[unstable(feature = "try_trait_v2", issue = "84277")]
 #[rustc_on_unimplemented(
     on(
-        all(from_method = "from_output", from_desugaring = "TryBlock"),
+        all(from_desugaring = "TryBlock"),
```
Comment (tmandry) [inline library/core/src/ops/try_trait.rs:118]: Why are the changes in this file needed?

File: src/test/ui/async-await/issues/issue-65436-raw-ptr-not-send.stderr (hunk 1 above)
Comment (estebank) [reply src/test/ui/async-await/issues/issue-65436-raw-ptr-not-send.stderr:12]: What output would be the desirable output here?

File: library/core/src/ops/try_trait.rs (hunk 4 above)
Comment (estebank) [reply library/core/src/ops/try_trait.rs:118]: I originally constrained these a lot to avoid false-positives, but given how the desugaring is applied, it wasn't necessary. But with the changes to the desugaring there were now some cases where we had `TryBlock`, but weren't coming from the method call (IIRC, we perform deduplication and a separate obligation ended up being selected for display). Removing the overspicificity fixed those regressions.

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs (hunk 3 above)
Comment (estebank) [reply compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:893]: I believe what I meant was to try and look at other cases beyond the following

```
error[E0277]: `()` is not a future
  --> $DIR/unnecessary-await.rs:9:10
   |
LL |     boo().await;
   |     -----^^^^^^ `()` is not a future
   |     |
   |     this call returns `()`
   |
   = help: the trait `Future` is not implemented for `()`
help: do not `.await` the expression
   |
LL -     boo().await;
LL +     boo();
   | 
help: alternatively, consider making `fn boo` asynchronous
   |
LL | async fn boo () {}
   | +++++
```

But if `boo` has a different return type and the call is `boo().await;` without an assignment, either change would potentially be incorrect. But I can't come up with a better example though :-/ I which I had been more detailed on my comment ^_^'. I can remove it.

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs [hunk 5]
```diff
@@ -873,6 +879,57 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
...
+        &self,
+        obligation: &PredicateObligation<'tcx>,
+        err: &mut DiagnosticBuilder<'_>,
+    ) {
+        let span = obligation.cause.span;
+
+        if let ObligationCauseCode::AwaitableExpr(hir_id) = obligation.cause.code {
+            err.span_suggestion_verbose(
+                span,
+                "do not `.await` the expression",
```
Comment (camelid) [inline compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:892]: This sounds a tiny bit blunt. What do you think of this, which to me sounds a bit more like other rustc suggestions?

```suggestion
                "remove the `.await`",
```

**Comment by tmandry** [comment]
r=me after rebase

**Comment by matthiaskrgr** [comment]
Hmm, I wonder if this failed this rollup: https://github.com/rust-lang/rust/pull/91775#issuecomment-991663377

It changes some stuff around async and the coverage test that fails shows some async stuff + assertion failure at the end ... 
But it's hard for me to look through all this so it's still quit a shot into the dark :/

@bors rollup=never 
please feel free to un-`rollup=never` if you think this is unrelated to the failure! 
````
 warning: 11 warnings emitted

thread 'main' panicked at 'attempt to add with overflow', ../coverage/overflow.rs:10:18
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: unused `Result` that must be used
  --> ../coverage/closure_macro_async.rs:44:5
   |
44 |     executor::block_on(test());
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_must_use)]` on by default
   = note: this `Result` may be an `Err` variant, which should be handled

warning: 1 warning emitted

thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `try and succeed`,
 right: `try and succeed`: this assert should fail', ../coverage/issue-84561.rs:123:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Error: ()
warning: unused imports: `future::Future`, `marker::Send`, `pin::Pin`
 --> ../coverage/async2.rs:4:5
  |
4 |     future::Future,
  |     ^^^^^^^^^^^^^^
5 |     marker::Send,
  |     ^^^^^^^^^^^^
6 |     pin::Pin,
  |     ^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

warning: 1 warning emitted

diff failed, and not suppressed without `// ignore-llvm-cov-show-diffs` in ../coverage/async.rs
make: *** [Makefile:123: async] Error 1
````

**Comment by tmandry** [comment]
Failed due to a timeout. I guess..

@bors r+ rollup=sometimes
